File: wear/src/main/java/com/pdyjak/powerampwear/music_browser/files/FilesBrowserFragment.java
package com.pdyjak.powerampwear.music_browser.files;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.android.gms.wearable.MessageEvent;
import com.pdyjak.powerampwear.MessageExchangeHelper;
import com.pdyjak.powerampwear.MessageListener;
import com.pdyjak.powerampwear.R;
import com.pdyjak.powerampwear.music_browser.BrowserFragmentBase;
import com.pdyjak.powerampwear.music_browser.Clickable;
import com.pdyjak.powerampwear.music_browser.ItemViewHolder;
import com.pdyjak.powerampwear.music_browser.MusicLibraryNavigator;
import com.pdyjak.powerampwear.music_browser.ViewHolderFactory;
import com.pdyjak.powerampwearcommon.requests.GetFilesRequest;
import com.pdyjak.powerampwearcommon.responses.File;
import com.pdyjak.powerampwearcommon.responses.FilesListResponse;
import com.pdyjak.powerampwearcommon.responses.Parent;

import java.util.ArrayList;
import java.util.List;

public class FilesBrowserFragment extends BrowserFragmentBase implements MessageListener {
    public static final String PARENT_KEY = "parent";

    private static class FilesViewHolderFactory implements ViewHolderFactory {
        @NonNull
        @Override
        public ItemViewHolder createViewHolder(ViewGroup parent) {
            View view = LayoutInflater.from(parent.getContext()).inflate(
                    R.layout.basic_two_line_item, parent, false
            );
            return new FileViewHolder(view);
        }
    }

    private Parent mParent;

    @NonNull
    @Override
    protected ViewHolderFactory createViewHolderFactory() {
        return new FilesViewHolderFactory();
    }

    @Override
    protected boolean tryRestoreCachedItems() {
        FilesListResponse cached = getMusicLibraryCache().getFilesList(mParent);
        if (cached == null) return false;
        setItems(transform(cached));
        return true;
    }

    @Override
    protected boolean shouldScrollTo(@NonNull Clickable item, @NonNull String scrollDest) {
        if (!(item instanceof FileItem)) return false;
        FileItem fileItem = (FileItem) item;
        return fileItem.title != null && fileItem.title.startsWith(scrollDest);
    }

    @Override
    protected void fetchItems() {
        GetFilesRequest request = new GetFilesRequest(mParent);
        MessageExchangeHelper helper = getMessageExchangeHelper();
        helper.addMessageListenerWeakly(this);
        helper.sendRequest(GetFilesRequest.PATH, request);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        ensureParent();
    }

    @Override
    protected void onGoingToRefresh() {
        ensureParent();
    }

    private void ensureParent() {
        Bundle args = getArguments();
        if (args == null) {
            throw new IllegalArgumentException("Should never happen");
        }
        mParent = args.getParcelable(PARENT_KEY);
    }

    @Override
    public void onPause() {
        super.onPause();
        getMessageExchangeHelper().removeMessageListener(this);
    }

    @Override
    public void onMessageReceived(@NonNull MessageEvent messageEvent) {
        if (FilesListResponse.PATH.equals(messageEvent.getPath())) {
            getMessageExchangeHelper().removeMessageListener(this);
            byte[] bytes = messageEvent.getData();
            if (bytes == null) return;
            FilesListResponse response = FilesListResponse.fromBytes(bytes);
            if (response != null) {
                setItems(transform(response));
                getMusicLibraryCache().update(response);
            }
        }
    }

    private List<Clickable> transform(@NonNull FilesListResponse response) {
        MusicLibraryNavigator navigator = getMusicLibraryNavigator();
        List<File> files = response.getFilesList();
        List<Clickable> items = new ArrayList<>();
        for (File file : files) {
            items.add(new FileItem(navigator, response.parent, file.id, file.title, file.artist,
                    file.album, file.duration, file.contextualId));
        }
        return items;
    }
}

File: wear/src/main/java/com/pdyjak/powerampwear/music_browser/files/FilesBrowserFragment.kt
package com.pdyjak.powerampwear.music_browser.files

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.google.android.gms.wearable.MessageEvent
import com.pdyjak.powerampwear.MessageListener
import com.pdyjak.powerampwear.R
import com.pdyjak.powerampwear.common.messageExchangeHelper
import com.pdyjak.powerampwear.common.musicLibraryCache
import com.pdyjak.powerampwear.common.musicLibraryNavigator
import com.pdyjak.powerampwear.music_browser.BrowserFragmentBase
import com.pdyjak.powerampwear.music_browser.Clickable
import com.pdyjak.powerampwear.music_browser.ItemViewHolder
import com.pdyjak.powerampwear.music_browser.ViewHolderFactory
import com.pdyjak.powerampwearcommon.requests.GetFilesRequest
import com.pdyjak.powerampwearcommon.responses.FilesListResponse
import com.pdyjak.powerampwearcommon.responses.Parent

class FilesBrowserFragment : BrowserFragmentBase(), MessageListener {

    companion object {
        const val PARENT_KEY = "parent"
    }

    private class FilesViewHolderFactory : ViewHolderFactory {
        override fun createViewHolder(parent: ViewGroup): ItemViewHolder {
            val view = LayoutInflater.from(parent.context).inflate(
                    R.layout.basic_two_line_item, parent, false
            )
            return FileViewHolder(view)
        }
    }

    private var mParent: Parent? = null

    override fun createViewHolderFactory(): ViewHolderFactory {
        return FilesViewHolderFactory()
    }

    override fun tryRestoreCachedItems(): Boolean {
        val cached = activity.musicLibraryCache.getFilesList(mParent) ?: return false
        setItems(transform(cached))
        return true
    }

    override fun shouldScrollTo(item: Clickable, scrollDest: String): Boolean {
        if (item !is FileItem) return false
        val fileItem = item
        return fileItem.title !== null && fileItem.title.startsWith(scrollDest)
    }

    override fun fetchItems() {
        val request = GetFilesRequest(mParent)
        val helper = activity.messageExchangeHelper
        helper.addMessageListenerWeakly(this)
        helper.sendRequest(GetFilesRequest.PATH, request)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        ensureParent()
    }

    override fun onGoingToRefresh() {
        ensureParent()
    }

    private fun ensureParent() {
        val args = arguments ?: throw IllegalArgumentException("Should never happen")
        mParent = args.getParcelable<Parent>(PARENT_KEY)
    }

    override fun onPause() {
        super.onPause()
        activity.messageExchangeHelper.removeMessageListener(this)
    }

    override fun onMessageReceived(messageEvent: MessageEvent) {
        if (FilesListResponse.PATH == messageEvent.path) {
            activity.messageExchangeHelper.removeMessageListener(this)
            val bytes = messageEvent.data ?: return
            val response = FilesListResponse.fromBytes(bytes)
            if (response !== null) {
                setItems(transform(response))
                activity.musicLibraryCache.update(response)
            }
        }
    }

    private fun transform(response: FilesListResponse): List<FileItem> {
        return response.filesList.map {
            FileItem(activity.musicLibraryNavigator, response.parent, it.id, it.title, it.artist,
                    it.album, it.duration, it.contextualId)
        }
    }
}
